File: src/test/java/kitchenpos/fake/FakeMenuRepository.java
package kitchenpos.fake;

import kitchenpos.domain.*;

import java.util.*;
import java.util.stream.Collectors;

public class FakeMenuRepository implements MenuRepository {

    private Map<UUID, Menu> menus = new HashMap<>();

    @Override
    public List<Menu> findAllByIdIn(List<UUID> ids) {
        return menus.keySet().stream()
                .filter(ids::contains)
                .map(id -> menus.get(id))
                .collect(Collectors.toList());
    }

    @Override
    public List<Menu> findAllByProductId(UUID productId) {
        return menus.values()
                .stream()
                .filter(menu -> hasProduct(menu.getMenuProducts(), productId))
                .collect(Collectors.toList());
    }

    private boolean hasProduct(List<MenuProduct> menuProducts, UUID productId) {
        return menuProducts.stream().anyMatch(it -> productId.equals(it.getProductId()));
    }

    @Override
    public Menu save(Menu menu) {
        menus.put(menu.getId(), menu);
        return menu;
    }

    @Override
    public Optional<Menu> findById(UUID id) {
        return Optional.ofNullable(menus.get(id));
    }

    @Override
    public List<Menu> findAll() {
        return menus.values()
                .stream()
                .collect(Collectors.toList());
    }
}

File: src/test/java/utils/FakeMenuRepository.kt
package utils

import kitchenpos.domain.Menu
import kitchenpos.domain.MenuRepository
import java.util.*

object FakeMenuRepository : MenuRepository {
    private val menuMap = mutableMapOf<UUID, Menu>()

    override fun save(menu: Menu): Menu {
        menuMap[menu.id] = menu

        return menu
    }

    override fun findById(id: UUID): Optional<Menu> {
        return Optional.ofNullable(menuMap[id])
    }

    override fun findAll(): List<Menu> {
        return menuMap.values.toList()
    }

    override fun findAllByIdIn(ids: List<UUID>): List<Menu> {
        return ids.asSequence()
            .mapNotNull { menuMap[it] }
            .toList()
    }

    override fun findAllByProductId(productId: UUID): List<Menu> {
        return menuMap.filterValues { menu ->
            menu.menuProducts.any { it.productId == productId }
        }.values.toList()
    }
}
